otter/test/convergence/test_service.py

```diff
@@ -1,95 +1,415 @@
-            '/groups/group-id/converge_lock')
-        # acquire is a monkey-patched partial function. :-(
-        self.kz_client.Lock().acquire.func.assert_called_once_with(timeout=120)
-        self.kz_client.Lock().release.assert_called_once_with()
-        expected_converge_args = (self.group, 0, self.lc, time(), log)
-        perform.assert_called_once_with(
-            self.dispatcher,
-            Effect(TenantScope(Effect(Constant(expected_converge_args)),
-                               'tenant-id')))
-
-    def test_converge_error_log(self):
-        """If performance fails, the error is logged."""
-        perform = mock.MagicMock()
-        perform.return_value = fail(ZeroDivisionError('foo'))
-        log = mock_log()
-        self.converger.start_convergence(
-            log,
-            self.group, self.state, self.lc,
-            execute_convergence=lambda *args: None,
-            perform=perform)
+        lock_set = make_lock_set()
+        my_buckets = [0, 5]
+        all_buckets = range(10)
+        result = converge_all_groups(
+            log, lock_set, my_buckets, all_buckets,
+            get_my_divergent_groups=get_my_divergent_groups,
+            converge_one_group=converge_one_group)
+        self.assertEqual(
+            sync_perform(_get_dispatcher(), result),
+            [('00', 'g1', 1, 'converge!'),
+             ('01', 'g2', 5, 'converge!')])
+        log.msg.assert_called_once_with(
+            'converge-all-groups',
+            group_infos=[{'tenant_id': '00', 'group_id': 'g1', 'version': 1},
+                         {'tenant_id': '01', 'group_id': 'g2', 'version': 5}])
 
-        log.err.assert_called_once_with(
-            CheckFailure(ZeroDivisionError),
-            "Error when performing convergence",
-            otter_msg_type='convergence-perform-error')
+
+class GetMyDivergentGroupsTests(SynchronousTestCase):
+
+    def test_get_my_divergent_groups(self):
+        """
+        :func:`get_my_divergent_groups` gets information about divergent groups
+        that are associated with the given buckets.
+        """
+        # sha1('00') % 10 is 6, sha1('01') % 10 is 1.
+        dispatcher = ComposedDispatcher([
+            EQDispatcher([
+                (GetChildrenWithStats(CONVERGENCE_DIRTY_DIR),
+                 [('00_gr1', ZNodeStatStub(version=0)),
+                  ('00_gr2', ZNodeStatStub(version=3)),
+                  ('01_gr3', ZNodeStatStub(version=5))]),
+            ]),
+            _get_dispatcher()
+        ])
+        result = sync_perform(
+            dispatcher, get_my_divergent_groups([6], range(10)))
+        self.assertEqual(
+            result,
+            [{'tenant_id': '00', 'group_id': 'gr1', 'version': 0},
+             {'tenant_id': '00', 'group_id': 'gr2', 'version': 3}])
+
+
+def _get_dispatcher():
+    return ComposedDispatcher([
+        TypeDispatcher({
+            ParallelEffects: perform_parallel_async,
```

Comment:
This is pretty useful. It will be nice if this `TypeDispatcher` can be there in `effect.async` for test purposes.